#!/usr/bin/env Rscript

###### 
# This script will pull out any GO terms that are significantly enriched for the 
# bacteria that were tested
# usage: analyze_GSEA_results_101614ERD.R <path.to.GSEA> <out.path> <q.val>
# 1. path.to.GSEA: path to GSEA files on cluster
# 2. out.path: path where the significant results should be written to
# 3. q.val: q-value cutoff to consider significance
# input: 	GSEA results files
# output: 	One table per bacteria and type of test listing any term that has q < 0.2
######

###### PARAMETERS ##########
# Set the parameters:
today <- Sys.Date()											# Set the date that will go on the end of the files generated by this script
today <- format(today, format="%m%d%y")
#############################



##### Read in arguments:
args <- commandArgs(trailingOnly=TRUE)
path.to.GSEA <- args[1]
out.path <- args[2]
q.val <- as.numeric(args[3])

#path.to.GSEA <- "/Users/erdavenport/clusterhome/poopQTL/14_seasons_individually/GSEA_results_winter_070114/"
#out.path <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/39_FC1to4_seasons_individually/9_GSEA_070114/winter/"
#q.val <- as.numeric("0.2")


##### Load GSEA files:
print("loading files")
GSEA.files <- list.files(path=paste(path.to.GSEA, sep=""))

bacteria <- unique(gsub("table.GSEA.enrichment.", "", gsub(".snps.*.from.nearest.gene.*.terms.*ERD.txt", "", GSEA.files)))

# Print to file any terms that have a significant q-value:
for (i in 1:length(bacteria)) {
	bacteria.files <- GSEA.files[grep(bacteria[i], GSEA.files)]
	BP <- read.table(file=paste(path.to.GSEA, bacteria.files[grep("BP", bacteria.files)], sep=""), sep="\t", header=TRUE, stringsAsFactors=FALSE, quote="")
	BP$type <- rep("BP", dim(BP)[1])
	
	CC <- read.table(file=paste(path.to.GSEA, bacteria.files[grep("CC", bacteria.files)], sep=""), sep="\t", header=TRUE, stringsAsFactors=FALSE, quote="")
	CC$type <- rep("CC", dim(CC)[1])
	
	MF <- read.table(file=paste(path.to.GSEA, bacteria.files[grep("MF", bacteria.files)], sep=""), sep="\t", header=TRUE, stringsAsFactors=FALSE, quote="")
	MF$type <- rep("MF", dim(MF)[1])
	
	types <- c("BP", "CC", "MF")
	mylist <- list(BP, CC, MF)

	for (j in 1:length(mylist)) {
		write.table(mylist[[j]][which(mylist[[j]][,4] < 0.2),], paste(out.path, "table.GSEA.q.",q.val,".",bacteria[i],".",types[j],".",today,"ERD.txt", sep=""), sep="\t", quote=FALSE)
	}
}

print("DONE!")