#!/usr/bin/env Rscript

###### 
# This script will generate a script to pull out SNP information to generate eQTL-like plots
# Usage:
# generate_script_to_pull_SNP_info_for_eQTL-like_plots_062514ERD.R <sig.bact.file> <input.path> <prefix> <out.path> <dropbox.out> <run.script>
# 1. sig.bact.file: path and file name of the table of names of files that contain significant GWAS SNPs
# 2. input.path: path to where the assoc files are located
# 3. prefix: prefix of the file name for the assoc files
# 4. out.path: the path on the cluster where you want the plink files written to
# 5. dropbox.out: the path on dropbox where you want files written to
# 6. run.script: the path to cluster where you want the plink run script to be written to
# input: 	list of bacteria with significant SNPs
#			.assoc.txt files generated from GEMMA for the bacteria associations
# output: 	script to run on cluster that will pull SNP information
#			table of bacteria to top SNP 
#			table of all significant rs numbers
######

###### PARAMETERS ##########
# Set the parameters:
today <- Sys.Date()											# Set the date that will go on the end of the files generated by this script
today <- format(today, format="%m%d%y")
#############################



##### Read command line arguments:
args <- commandArgs(trailingOnly=TRUE)
sig.bact.file <- args[1] # File that lists the names of the files that contain significant GWAS SNPs
input.path <- args[2] # Path to where the assoc files with Q values live
prefix <- args[3] # prefix on assoc file
out.path <- args[4] # outpath where you want plink files written to
dropbox.out <- args[5] # outpath where you want files written to on dropbox
run.script <- args[6] # outpath on cluster where you want the script to run plink written to

print(paste("file of significant bacteria: ", sig.bact.file, sep=""))
print(paste("path to association files with q-values: ", input.path, sep=""))
print(paste("prefix of assoc files: ", prefix, sep=""))
print(paste("path where plink files are written: ", out.path, sep=""))
print(paste("path to dropbox files: ", dropbox.out, sep=""))
print(paste("run script for plink: ", run.script, sep=""))

# sig.bact.file <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/39_FC1to4_seasons_individually/3_GEMMA_for_bacteria_052614/winter/traits.all.significant.052614ERD.txt"
# input.path <- "/Users/erdavenport/clusterhome/poopQTL/14_seasons_individually/GEMMA_output_bacteria_winter_qvals_052614/"
# prefix <- "hutt.3chip.hg19remap.winter.GWAS."
# out.path <- "/mnt/lustre/home/erdavenport/poopQTL/14_seasons_averaged/GEMMA_output_winter_bacteria_qvals_051414/eQTL_plot_files_052614/"
# dropbox.out <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/39_FC1to4_seasons_individually/3_GEMMA_for_bacteria_052614/winter/"
# run.script <- "/Users/erdavenport/clusterhome/Programs/Gemma/recode_SNPs_for_eQTL_plots_062514ERD.sh"


##### Load in the table of significant bacteria:
sig.bact <- read.table(file=sig.bact.file, sep="\t", header=FALSE)$V1

filez <- list.files(input.path, pattern=".assoc.")

##### Generate script to pull out relavant bacteria on cluster:
print("writing scripts to pull SNP information")
all.sig.snps <- c()
script <- c("#!/bin/bash")
for (i in 1:length(sig.bact)) {
	bacteria <- gsub(prefix, "", gsub(".column.*.txt", "", sig.bact[i]))
	assoc <- read.table(file=paste(input.path,filez[grep(bacteria, filez)], sep=""), sep=" ", header=TRUE, colClasses=c("numeric", "character", "numeric", "numeric", "character", "character", "numeric", "numeric", "numeric"))

	snps <- assoc[which(assoc$p_lrt <= 0.05/dim(assoc)[1] | assoc$q_lrt <= 0.2),]
	snps <- cbind(rep(bacteria, dim(snps)[1]), snps)
	all.sig.snps <- rbind(all.sig.snps, snps)
	

	# recode SNPs:
	for (i in 1:dim(snps)[1]) {
		script <- c(script, paste("~/Programs/plink/plink --bfile /mnt/lustre/home/erdavenport/Programs/Gemma/hutt.3chip.hg19remap --snp ", snps$rs[i]," --recode --noweb --out ",out.path,"recode_",snps$rs[i],"_",bacteria,"_",today,"ERD.txt", sep=""))
		# pull out major and minor alleles:
		script <- c(script, paste('grep "', snps$rs[i],'" /mnt/lustre/home/erdavenport/Programs/Gemma/hutt.3chip.hg19remap.bim >> ',out.path,'major_minor_alleles_for_eQTL_plots_',today,'ERD.txt',sep=""))
	}
}

colnames(all.sig.snps)[1] <- "bacteria"

write.table(all.sig.snps$rs, paste(dropbox.out,"all_sig_SNPs_from_bacteria_GWAS_rs_only_", today, "ERD.txt", sep=""), sep="\n", row.names=FALSE, col.names=FALSE, quote=FALSE)
write.table(all.sig.snps, paste(dropbox.out,"all_sig_SNPs_from_bacteria_GWAS_",today,"ERD.txt", sep=""), sep="\t", row.names=FALSE, quote=FALSE)
write.table(script, run.script, sep="\n", row.names=FALSE, col.names=FALSE, quote=FALSE)

print("DONE!")