#!/usr/bin/env Rscript

###### 
# This script generates a shell script that will submit jobs on the cluster to run GSEA
# on any bacteria that has a significant SNP p-value in any of the GWAS examined.
# Usage: ./generate_GSEA_submission_script_sig_only_070114ERD.R <sig.table> <KB> <prefix> <cluster.loc> <out.path> <out.script>
# 1. sig.table: path and name of the file that lists all of the significant GWAS files
# 2. KB: distance away from genes (in kb) that you want to consider for the GSEA  
# 3. prefix: prefix of the gemma assoc file
# 4. cluster.loc: path on cluster to assoc files 
# 5. out.path: path where you want result tables to be written
# 6. out.script: name and location where the run script should be written to
# input: 	table of bacteria with a Bonferroni sig p-value in either winter, summer, or seasons combined
# output: 	shell script to run on the cluster that will submit the GSEA script
######

###### PARAMETERS ##########
# Set the parameters:
today <- Sys.Date()											# Set the date that will go on the end of the files generated by this script
today <- format(today, format="%m%d%y")
#############################



##### Read in command line arguments:
args <- commandArgs(trailingOnly=TRUE)
sig.table <- args[1] # list of all significant associations examined
KB <- args[2] # distance away from genes (in kb) that you want to consider for the GSEA
prefix <- args[3] # prefix of the GWAS assoc file
cluster.loc <- args[4] # path on cluster to where the gwas files are located
out.path <- args[5] # path to where you want the output to be written
out.script <- args[6] # name and location where the run script should be written to

# sig.table <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/39_FC1to4_seasons_individually/3_GEMMA_for_bacteria_052614/winter/traits.all.significant.052614ERD.txt"
# KB <- 10
# prefix <- "hutt.3chip.hg19remap.winter.GWAS."
# cluster.loc <- "/mnt/lustre/home/erdavenport/poopQTL/14_seasons_individually/GEMMA_output_bacteria_winter_052614/"
# out.path <- "/mnt/lustre/home/erdavenport/poopQTL/14_seasons_individually/GSEA_results_winter_070114/"
# out.script <- "/Users/erdavenport/clusterhome/repos/scripts_repo/poop_scripts/Non-general_scripts/submit_GSEA_significant_bacteria_winter_"

##### Load in significant bacteria for each seasonal combination:
print("loading in data")
sigs <- read.table(file=sig.table, sep="\n", header=FALSE)$V1



##### Write submission script for doing GSEA:
print("writing script")
script <- c("#!/bin/bash")
for (i in 1:length(sigs)) {
	bacteria <- gsub(prefix, "", gsub(".column.*.assoc.txt", "", sigs[i]))
	script <- c(script, paste("qsub -l h_vmem=2g -V -b y ~/Programs/program_files/bin/Rscript ~/repos/scripts_repo/poop_scripts/GSEA_postgwas_with_distance_trimming_040414ERD.R ",cluster.loc,sigs[i]," ", bacteria," ",KB," ", out.path, sep=""))
}



##### Write out the table:
print("saving script to the cluster")
write.table(script, paste(out.script, today,"ERD.sh", sep=""), sep="\n", row.names=FALSE, col.names=FALSE, quote=FALSE)


print("DONE!")