#!/usr/bin/env Rscript

###### 
# This script will take the initial bacterial proportions file and trim it down to 
# the number of samples you want to look at. It will then also ensure that the sample order
# is the same in the phenotype files (for the traits and diseases) and covariates files.
# This particular version with combine the counts for all winter and summer as if they
# are technical replicates, not considering DE at all. 
# This will make the diversity table for seasons added. 
# input: 	data tables of bacteria (5 levels)
# output: 	diversity table with normalized and unnormalized diversity metrics
######

###### PARAMETERS ##########
# Set the parameters:
today <- Sys.Date()							# Set the date that will go on the end of the files generated by this script
today <- format(today, format="%m%d%y")
output.path <- c("/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/38_FC1to4_seasons_added/")	# path for all output to go into
subfolder <- c("8_diversity_061914/")		# Optional folder for putting initial data files into
#############################


##### Load libraries:
library(vegan)



##### First, load in all levels of data across all individuals, both seasons:
all.bacteria <- c()
print("Loading genus proportion tables...")
for (a in c("genus")) {
	data <- read.table(file=paste("/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/20_FC1to4_for_QTLs/FC1to4.subsampled.2mil.",a,".counts.reps.combined.no.abs.052313ERD.txt", sep=""), sep="\t", header=TRUE)
	print(c(a, dim(data)[1]))
	all.bacteria <- rbind(all.bacteria, data)
}
print(c("all taxa", dim(all.bacteria)[1]))



##### Trim to just the individuals we're interested in looking at - eliminate the individual without genotype data:
print("trimming individuals from full table:")
all.bacteria <- all.bacteria[,-grep("154112", colnames(all.bacteria))]



##### Combine winter and summer counts:
# Separate into winter and summer to prune individuals:
w <- all.bacteria[,grep("W", colnames(all.bacteria))]
s <- all.bacteria[,grep("S", colnames(all.bacteria))]

# Which individuals are in both seasons?
both <- gsub("W", "", colnames(w)[which(gsub("W", "", colnames(w)) %in% gsub("S", "", colnames(s)))])
winter <- gsub("W", "", colnames(w)[which(!gsub("W", "", colnames(w)) %in% gsub("S", "", colnames(s)))])
summer <- gsub("S", "", colnames(s)[which(!gsub("S", "", colnames(s)) %in% gsub("W", "", colnames(w)))])

# make a table of the winter and summer samples that weren't in both seasons:
combined <- w[,which(gsub("W", "", colnames(w)) %in% winter)]

season <- c(rep("winter", dim(combined)[2]))
combined <- cbind(combined, s[,which(gsub("S", "", colnames(s)) %in% summer)])
season <- c(season, rep("summer", length(which(gsub("S", "", colnames(s)) %in% summer))))

# add the summer and winter counts for samples that were in both seasons:
w <- w[,which(gsub("W", "", colnames(w)) %in% both)]
s <- s[,which(gsub("S", "", colnames(s)) %in% both)]

ws <- w + s
season <- c(season, rep("both", dim(ws)[2]))

# Create one table of "combined" taxa:
combined <- cbind(combined, ws)
colnames(combined) <- gsub("_[WS]", "", colnames(combined))



##### Standardize tables by total reads after subsampling:
reads <- read.table(file=paste("/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/20_FC1to4_for_QTLs/FC1to4.subsampled.taxonomy.lines.reps.combined.no.abs.052313ERD.txt", sep=""), sep="\t", header=TRUE)

# Combined table:
w.reads <- reads[grep("W", reads$V2),]
s.reads <- reads[grep("S", reads$V2),]

w.reads$V3 <- gsub("W", "", w.reads$V2)
s.reads$V3 <- gsub("S", "", s.reads$V2)

ws.reads <- merge(w.reads, s.reads, by="V3", all=TRUE)
ws.reads[is.na(ws.reads)] <- 0
ws.reads$total <- ws.reads$V1.x + ws.reads$V1.y

for (i in 1:dim(combined)[2]) {
	combined[,i] <- combined[,i]/ws.reads$total[which(gsub("_$", "", ws.reads$V3) == colnames(combined)[i])]
}



##### Calculate diversity easures for each person:
print("calculating diversity metrics")
S <- specnumber(t(combined))
H <- diversity(t(combined), index="shannon")
J <- H/log(S)

diversity <- cbind(S, H, J)



##### qqnorm the diversity metrics (in addition to leaving them as is):
normed.log.QTL.data <-  diversity

set.seed(1)	
for (i in 1:dim(normed.log.QTL.data)[2]) {
	x <- sample(1:dim(normed.log.QTL.data)[1])
	normed <- qqnorm(normed.log.QTL.data[x,i], plot.it=FALSE)$x
	normed.log.QTL.data[,i] <- normed[order(x)]
}	

colnames(normed.log.QTL.data) <- paste("normalized_",colnames(normed.log.QTL.data), sep="")

log.QTL.data <- t(cbind(diversity, normed.log.QTL.data))



##### Write out diversity tables:
print("writing out diversity tables")
write.table(log.QTL.data, paste(output.path, subfolder, "diversity.measures.added.",today,"ERD.txt", sep=""), sep="\t", quote=FALSE)

print("DONE!")
