#!/usr/bin/env Rscript

###### 
# This script will plot Maurano/Duke enrichments for all single cell types that have at least
# one tissue in the last bin with a FET p-value < 0.05.
# Usage:
# ./plot_Maurano_enrichment_plots_single_cell_type_single_panels_FET_063014.R <tissue.file> <base.path> <enr.path> <p.path> <plot.path>
# 1. tissue.rile: File from Darren with the tissue classification data
# 2. base.path: root directory to where all of the Maurano analysis files are stored
# 3. enr.path: folder where enrichment tables are stored (from base path)
# 4. p.path: path to where FET p-values are stored (from base path)
# 5. plot.path: path to where pdfs of shotgun plots should be written (from base path)
# input: 		enrichment tables from maurano_enrichment_script
#				pval table from maurano_enrichment_script
#				cellTypes table from Encode	
# output: 		plot of enrichments by cell type for bacteria with at least one significant
#				cell type
######

###### PARAMETERS ##########
# Set the parameters:
today <- Sys.Date()											# Set the date that will go on the end of the files generated by this script
today <- format(today, format="%m%d%y")
#############################



##### Read in command line arguments:
args <- commandArgs(trailingOnly=TRUE)
tissue.file <- args[1] # tissue info file
base.path <- args[2] # base path to Maurano folder
enr.path <- args[3] # path to where the enrichments are
p.path <- args[4] # path to where the FET p-values are
plot.path <- args[5] # path to where the plot files should be written

# tissue.file <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/37_FC1to4_seasons_averaged/6_Maurano_enrichment_051514/stam_cells.txt"
# base.path <- "/Users/erdavenport/Dropbox/Lab/Poop/Sequencing_Results/39_FC1to4_seasons_individually/6_Maurano_enrichment_052814/summer/"
# enr.path <- "1_enrichments_052814/"
# p.path <- "3_FET_pvals_062614/"
# plot.path <- "4_single_plots_FET_063014/"


 
 
##### Load celltype data:
print("loading in cell type, enrichment, and p-value data")
celltypes <- read.table(file=tissue.file, sep="\t", header=FALSE, stringsAsFactors=FALSE)
celltypes[is.na(celltypes)] <- "other"


##### Load enrichment data:
enri <- list.files(path=paste(base.path, enr.path, sep=""))


##### Load the p-values:
pvals <- list.files(path=paste(base.path, p.path, sep=""))


##### Set color scheme:
##### Define cell type colors:
plot.col <- function(x) {
	if (x == "other") {
		return("gray")
	} else if (x == "Adrenal") {
		return("orange")
	} else if (x == "Brain") {
		return("hotpink")
	} else if (x == "Cancer") {
		return("purple") 
	} else if (x == "Endothelial") {
		return("red4")
	} else if (x == "Heart") {
		return("red")
	} else if (x == "Intestine") {
		return("chocolate")
	} else if (x == "Kidney") {
		return("darkgoldenrod1")
	} else if (x == "Lung") {
		return("deepskyblue")
	} else if (x == "Lymphoid") {
		return("navajowhite3")
	} else if (x == "Muscle") {
		return("burlywood4")
	} else if (x == "Myeloid") {
		return("magenta")
	} else if (x == "Placenta") {
		return("mistyrose4") 
	} else if (x == "Skin") {
		return("seagreen4") 
	} else if (x == "StemCell") {
		return("yellowgreen")
	} else if (x == "Stomach") {
		return("tan3")
	} else if (x == "Thymus") {
		return("royalblue")
	}
}

ts <- names(table(celltypes$V2))
ts <- ts[-which(ts == "other")]

print("plotting enrichments")

for (i in 1:length(enri)) {
	bacteria <- gsub("table.Maurano.enrichments.", "", gsub("[.][0-9]*ERD.txt", "", enri[i]))
	enrichments <- read.table(file=paste(base.path, enr.path,enri[grep(bacteria, enri)], sep=""), sep="\t", header=TRUE, stringsAsFactors=FALSE)
	pvalues <- read.table(file=paste(base.path, p.path, pvals[grep(bacteria,enri)], sep=""), sep="\t", header=TRUE, stringsAsFactors=FALSE)
	rownames(pvalues) <- gsub(".hg19", "", rownames(pvalues))
	new.thresholds <- gsub("X", "", colnames(enrichments))	
	
	# Only if there's at least one tissue with a FET p-value < 0.05 in the lowest bin will we generate a plot:
	if (min(pvalues[,dim(pvalues)[2]]) <= 0.05) {
		pdf(paste(base.path, plot.path, "plot.Maurano.enrichments.single.track.pvals.",bacteria,".",today,"ERD.pdf", sep=""), width=8, height=12)
		par(mfrow=c(3, 2))
		
		# Rename cell types so you're only looking at the one you want to look at and other:
		for (j in 1:length(ts)) {
			# Make sure the order of the cell types from Darren are the same as in the enrichments and pvalues table:
			myfun <- function(x) {which(rownames(pvalues) == x)}
			celltypes2 <- sapply(celltypes$V1, myfun)
			cell.types <- celltypes[order(celltypes2),2]
			
			# Only consider the cell type of interest:
			cell.types[which(cell.types != ts[j])] <- "other"
	
			# plot it up:
			ctype <- cell.types[1]
			plot(1:length(new.thresholds), enrichments[1,], type="l", lwd=2, col=plot.col(ctype), ylim=c(0,max(enrichments)), xaxt="n", xlab="GWAS p-value threshold", ylab="Fold enrichment of SNPs in DHSs", main=bacteria)
			axis(1, at=1:length(new.thresholds), labels=new.thresholds, las=2)
			# First plot the gray lines:
			for (m in 1:dim(enrichments)[1]) {
				if (cell.types[m] == "other") {
					ctype <- cell.types[m]
					points(1:length(new.thresholds), enrichments[m,], type="l", lwd=2, col=plot.col(ctype), xaxt="n")
				}
			}
	
			# Then plot relevant cell types:
			for (m in 1:dim(enrichments)[1]) {
				if (cell.types[m] != "other") {
					ctype <- cell.types[m]
					points(1:length(new.thresholds), enrichments[m,], type="l", lwd=2, col=plot.col(ctype), xaxt="n")
				}
			}
			
			# Add significance measures to plot:
			for (m in 1:dim(enrichments)[1]) {
				ctype <- cell.types[m]
				if (ctype != "other") {
					if (pvalues[m, dim(pvalues)[2]] <= 0.05) {
						text(length(new.thresholds), enrichments[m,length(new.thresholds)], adj = 0, labels = "*", cex=2, col=plot.col(ctype))
					}
				} 
			}
	
			# Set up legend:
			cells <- unique(cell.types)
			cells2 <- c()
			for (j in 1:length(cells)) {
				cells2 <- c(cells2, paste(cells[j], " (n=", length(which(cell.types == cells[j])), ")", sep=""))
			}
			legend("topleft", fill=sapply(unique(cell.types), plot.col), legend=cells2, bty="n")
		}	
		dev.off()
	}
}

print("DONE!")